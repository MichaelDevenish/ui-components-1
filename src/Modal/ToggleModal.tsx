import React from 'react'
import { Button, IButtonProps } from '../Button'
import { Modal, IModalProps } from './Modal'

interface IToggleArguments {
  toggle: () => void
}

interface IToggleModalProps {
  /** A trigger component which opens or closes the modal */
  trigger: (toggleArguments: IToggleArguments) => JSX.Element

  /** The children inside the Modal component */
  children: JSX.Element
}

interface IToggleModalState {
  isOpen: boolean
}

class ToggleModal extends React.PureComponent<IToggleModalProps & IModalProps, IToggleModalState> {
  public state = {
    isOpen: false
  }

  public render (): JSX.Element {
    const {
      isOpen
    } = this.state

    const {
      trigger
    } = this.props

    return (
      <React.Fragment>
        {trigger({toggle: this.toggle})}
        <Modal
          isOpen={isOpen}
          handleClose={this.toggle}
          {...this.props}
        />
      </React.Fragment>
    )
  }

  private toggle = () => {
    const {
      isOpen
    } = this.state

    this.setState({ isOpen: !isOpen })
  }
}

export {
  IToggleArguments,
  IToggleModalProps,
  IToggleModalState,
  ToggleModal
}
