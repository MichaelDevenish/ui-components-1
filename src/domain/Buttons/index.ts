export { BaseButton, IBaseButtonProps } from './BaseButton'
export { LinkButton } from './LinkButton'
export { Button } from './Button'
export { ButtonGroup } from './ButtonGroup'
export { FontAwesomeIconButton } from './FontAwesomeIconButton'
