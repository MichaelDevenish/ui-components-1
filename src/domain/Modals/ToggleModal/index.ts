export { IToggleArguments, IToggleModalProps, IToggleModalState, ToggleModal } from './ToggleModal'
