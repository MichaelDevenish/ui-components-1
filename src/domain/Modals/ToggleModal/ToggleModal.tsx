import React from 'react'
import { Modal, IBaseModalProps } from '../Modal'

interface IToggleArguments {
  toggle: () => void
}

interface IToggleModalProps extends IBaseModalProps {
  /** A trigger component which opens or closes the modal */
  trigger: (toggleArguments: IToggleArguments) => JSX.Element
}

interface IToggleModalState {
  isOpen: boolean
}

class ToggleModal extends React.PureComponent<IToggleModalProps, IToggleModalState> {
  public state = {
    isOpen: false
  }

  public render (): JSX.Element {
    const {
      isOpen
    } = this.state

    const {
      trigger,
      ...props
    } = this.props

    return (
      <React.Fragment>
        {trigger({toggle: this.toggle})}
        <Modal
          isOpen={isOpen}
          handleClose={this.toggle}
          {...props}
        />
      </React.Fragment>
    )
  }

  private toggle = () => {
    const {
      isOpen
    } = this.state

    this.setState({ isOpen: !isOpen })
  }
}

export {
  IToggleArguments,
  IToggleModalProps,
  IToggleModalState,
  ToggleModal
}
