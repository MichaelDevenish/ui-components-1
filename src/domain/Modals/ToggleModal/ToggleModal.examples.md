### Toggle Modal

Toggle Modal provides you a basic toggle function to
show or hide a modal without the need to implement
by yourself.

#### Toggle Modal with Button

```jsx
import { Button } from '@Domain/Buttons';

<ToggleModal
  trigger={({toggle}) => (
    <Button
      type='primary'
      onClick={toggle}
    >
      Click Me
    </Button>
  )}
>
  Hello this is a Modal
</ToggleModal>
```

#### Toggle Modal with Anchor

```jsx
import { Anchor } from '@Domain/Internals';

<ToggleModal
  trigger={({toggle}) => (
    <Anchor
      href='#/UI Components/Modals/ToggleModal'
      onClick={toggle}
    >
      Click Me
    </Anchor>
  )}
>
  Hello this is a Modal
</ToggleModal>
```
