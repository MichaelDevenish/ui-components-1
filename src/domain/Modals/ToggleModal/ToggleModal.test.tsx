import { mount, shallow } from 'enzyme'
import React from 'react'

import { Button } from '../../Buttons'
import { ToggleModal } from './ToggleModal'

describe('<ToggleModal />', () => {
  const trigger = ({toggle}: any) => (
    <Button
      type='primary'
      onClick={toggle}
    >
      Click Me
    </Button>
  )

  it('should render a modal with a button', () => {
    const wrapper = shallow(
      <ToggleModal
        trigger={trigger}
      >
        Hello this is a Modal
      </ToggleModal>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should open a modal after click', () => {
    const wrapper = mount(
      <ToggleModal
        trigger={trigger}
      >
        <div>hello this is a modal</div>
      </ToggleModal>
    )

    wrapper.find('button').simulate('click')

    expect(wrapper.find('div[children="hello this is a modal"]').exists()).toBeTruthy()
  })
})
