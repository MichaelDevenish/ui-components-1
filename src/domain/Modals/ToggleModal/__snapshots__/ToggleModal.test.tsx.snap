// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleModal /> should render a modal with a button 1`] = `
<Fragment>
  <Button
    hasLegacyMargins={false}
    iconAlignment="left"
    onClick={[Function]}
    type="primary"
  >
    Click Me
  </Button>
  <Modal
    handleClose={[Function]}
    isOpen={false}
    offsetZIndex={0}
    showCloseButton={true}
    size="medium"
    useSubcomponents={false}
  >
    Hello this is a Modal
  </Modal>
</Fragment>
`;
